// ngx-disqus/bundles/ngx-disqus.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core'], factory) :
	(factory((global['ngx-disqus'] = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SHORTNAME = new core.InjectionToken('SHORTNAME');
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var DisqusService = (function () {
    /**
     * @param {?} shortname
     */
    function DisqusService(shortname) {
        this.shortname = shortname;
    }
    Object.defineProperty(DisqusService.prototype, "window", {
        /**
         * @return {?}
         */
        get: function () {
            return _window();
        },
        enumerable: true,
        configurable: true
    });
    return DisqusService;
}());
DisqusService.decorators = [
    { type: core.Injectable },
];
/** @nocollapse */
DisqusService.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core.Inject, args: [SHORTNAME,] },] },
]; };
/**
 * @return {?}
 */
function _window() {
    return typeof window !== 'undefined' ? window : global;
}
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var DisqusComponent = (function () {
    /**
     * @param {?} renderer
     * @param {?} el
     * @param {?} dService
     */
    function DisqusComponent(renderer, el, dService) {
        this.renderer = renderer;
        this.el = el;
        this.dService = dService;
        /**
         * DISQUS events
         */
        this.newComment = new core.EventEmitter(true);
        this.ready = new core.EventEmitter(true);
        this.paginate = new core.EventEmitter(true);
    }
    /**
     * @return {?}
     */
    DisqusComponent.prototype.ngOnChanges = function () {
        /** Reset Disqus if any input changed */
        if (!this.dService.window.DISQUS) {
            this.addDisqusScript();
        }
        else {
            this.reset();
        }
    };
    /**
     * Add DISQUS script
     * @return {?}
     */
    DisqusComponent.prototype.addDisqusScript = function () {
        /** Set DISQUS config */
        this.dService.window.disqus_config = this.getConfig();
        var /** @type {?} */ disqusScript = this.renderer.createElement('script');
        disqusScript.src = "//" + this.dService.shortname + ".disqus.com/embed.js";
        disqusScript.async = true;
        disqusScript.type = 'text/javascript';
        this.renderer.setAttribute(disqusScript, 'data-timestamp', new Date().getTime().toString());
        this.renderer.appendChild(this.el.nativeElement, disqusScript);
    };
    /**
     * Reset DISQUS with the new config
     * @return {?}
     */
    DisqusComponent.prototype.reset = function () {
        this.dService.window.DISQUS.reset({
            reload: true,
            config: this.getConfig()
        });
    };
    /**
     * Create DISQUS config from inputs
     * @return {?}
     */
    DisqusComponent.prototype.getConfig = function () {
        var /** @type {?} */ self = this;
        return function () {
            this.page.identifier = self.identifier;
            this.page.url = self.validateUrl(self.url);
            this.page.title = self.title;
            this.category_id = self.category;
            this.language = self.language;
            /* Available callbacks are afterRender, onInit, onNewComment, onPaginate, onReady, preData, preInit, preReset */
            this.callbacks.onNewComment = [function (e) {
                    self.newComment.emit(e);
                }];
            this.callbacks.onReady = [function (e) {
                    self.ready.emit(e);
                }];
            this.callbacks.onPaginate = [function (e) {
                    self.paginate.emit(e);
                }];
        };
    };
    /**
     * @param {?} url
     * @return {?}
     */
    DisqusComponent.prototype.validateUrl = function (url) {
        /** Validate URL input */
        if (url) {
            var /** @type {?} */ r = /(http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/;
            if (r.test(url)) {
                return url;
            }
            else {
                console.warn('[Disqus]: Invalid URL, return undefined');
            }
        }
        /** DISQUS will fallback to "Window.location.href" when URL is undefined */
        return undefined;
    };
    /**
     * @return {?}
     */
    DisqusComponent.prototype.ngOnDestroy = function () {
        this.dService.window.DISQUS = undefined;
        this.dService.window.disqus_config = undefined;
    };
    return DisqusComponent;
}());
DisqusComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'disqus',
                template: '<div id="disqus_thread"></div>',
                changeDetection: core.ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
DisqusComponent.ctorParameters = function () { return [
    { type: core.Renderer2, },
    { type: core.ElementRef, },
    { type: DisqusService, },
]; };
DisqusComponent.propDecorators = {
    "url": [{ type: core.Input },],
    "identifier": [{ type: core.Input },],
    "title": [{ type: core.Input },],
    "category": [{ type: core.Input },],
    "language": [{ type: core.Input },],
    "newComment": [{ type: core.Output },],
    "ready": [{ type: core.Output },],
    "paginate": [{ type: core.Output },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Initialize Disqus with shortname
 * @param {?} shortname
 * @return {?}
 */
function DisqusFactory(shortname) {
    return new DisqusService(shortname);
}
var DisqusModule = (function () {
    function DisqusModule() {
    }
    /**
     * @param {?} shortname
     * @return {?}
     */
    DisqusModule.forRoot = function (shortname) {
        return {
            ngModule: DisqusModule,
            providers: [
                { provide: SHORTNAME, useValue: shortname },
                {
                    provide: DisqusService,
                    useFactory: DisqusFactory,
                    deps: [SHORTNAME]
                }
            ]
        };
    };
    return DisqusModule;
}());
DisqusModule.decorators = [
    { type: core.NgModule, args: [{
                declarations: [DisqusComponent],
                exports: [DisqusComponent]
            },] },
];
/** @nocollapse */
DisqusModule.ctorParameters = function () { return []; };

exports.DisqusModule = DisqusModule;
exports.ɵb = DisqusComponent;
exports.ɵa = DisqusFactory;
exports.ɵc = DisqusService;
exports.ɵd = SHORTNAME;

Object.defineProperty(exports, '__esModule', { value: true });

})));
